### PR TITLE
fix(typing) cleanup snuba.datasets

### PR DESCRIPTION
**Pull request by evanh**

### CODE REVIEW

File: snuba/cli/subscriptions.py [hunk 1]
```diff
@@ -119,6 +119,7 @@ def subscriptions(
     dataset = get_dataset(dataset_name)
 
     storage = dataset.get_default_entity().get_writable_storage()
+    assert storage is not None
```
Comment (fpacifici) [inline snuba/cli/subscriptions.py:122]: Please provide a meaningful error message as the user will see this and won't know why their input was accepted but subscriptions are not working.

File: snuba/datasets/entity.py [hunk 2]
```diff
@@ -222,7 +222,7 @@ def get_function_call_validators(self) -> Mapping[str, FunctionCallValidator]:
         """
         return {}
 
-    def get_writable_storage(self) -> WritableTableStorage:
+    def get_writable_storage(self) -> Optional[WritableTableStorage]:
```
Comment (fpacifici) [inline snuba/datasets/entity.py:225]: If the issue is only outcomes_raw. Why cannot we add the writable storage there ?

Comment (lynnagara) [reply snuba/datasets/entity.py:225]: Also discover

File: snuba/datasets/table_storage.py [hunk 3]
```diff
@@ -170,8 +170,8 @@ def get_schema(self) -> WritableTableSchema:
     def get_batch_writer(
         self,
         metrics: MetricsBackend,
-        options=None,
-        table_name=None,
+        options: Optional[ClickhouseWriterOptions] = None,
```
Comment (lynnagara) [inline snuba/datasets/table_storage.py:173]: `ClickhouseWriterOptions` is already optional

**Review by fpacifici — APPROVED** [review]

**Pull request merged**